### PR TITLE
bpo-44732: Rename types.Union to types.UnionType

### PR DESCRIPTION
**Pull request by AliyevH**
https://bugs.python.org/issue44732

### CODE REVIEW

**Review by FFY00 — CHANGES_REQUESTED** [review]
A plain rename from `Union` to `UnionType` would be a big breaking change, it should be deprecated IMO.

We could do that in `types.__getattr__`, by adding the following to the `types` module.

```python
import warnings

...

def __getattr__(name):
    if name == 'Union':
        warnings.warn('types.Union is deprecated and should be replaced with types.UnionType, it will get removed in X time', DeprecationWarning)
        return UnionType
    raise AttributeError(f'module {__name__} has no attribute {name}')
```

File: Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst [hunk 1]
```diff
@@ -1,2 +1,2 @@
-Add ability to serialise ``types.Union`` objects. Patch provided by Yurii
+Add ability to serialise ``types.UnionType`` objects. Patch provided by Yurii
```
Comment (FFY00) [inline Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: We should not change this news entry :stuck_out_tongue:

Comment (AliyevH) [reply Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: Didn't know about it 😕

Comment (FFY00) [reply Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: Oh wait, maybe we do actually want this changed as it hasn't been released yet :facepalm:. Sorry!

Comment (AliyevH) [reply Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: no problem.
Can you please tell me X time in deprecation warning?😬

Comment (FFY00) [reply Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: I don't know, that's up to the core devs, ask in https://bugs.python.org/issue44732 if we should have a deprecation period and if so, how long.

Comment (AliyevH) [reply Misc/NEWS.d/next/Core and Builtins/2021-07-19-19-53-46.bpo-44676.WgIMvh.rst]: Okey thanks 😄

**Comment by Fidget-Spinner** [comment]
@FFY00 

> A plain rename from `Union` to `UnionType` would be a big breaking change, it should be deprecated IMO.

From what I understand, it won't be. It looks like Serhiy is planning to backport the changes to 3.10 too where Union was first introduced. Please wait a bit to see what he has to say about the PR. Thanks!

**Comment by AliyevH** [comment]
Thanks @Fidget-Spinner . Okey let's wait 😃

**Review by serhiy-storchaka — APPROVED** [review]
LGTM!

**Comment by JelleZijlstra** [comment]
I don't see a NEWS entry in the PR but bedevere passed, am I missing something?

I do think there should be a NEWS entry for this.

**Comment by gvanrossum** [comment]
> I don't see a NEWS entry in the PR but bedevere passed, am I missing something?
> 
> I do think there should be a NEWS entry for this.

The OP updated all of the existing NEWS entries that mention types.Union. That's creative, but I don't think it should be done that way. We need a new NEWS entry describing the renaming.

**Comment by AliyevH** [comment]
As i understand, we don't touch news files and add a new one with a description of renaming types.Union to types.UnionType.
I will update this today.

File: Lib/types.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 Define names for built-in types that aren't directly accessible as a builtin.
 """
 import sys
+import warnings
```
Comment (uriyyo) [inline Lib/types.py]: Do we need import `warnings`?  I can not see the usage of this module at `types.py`.

Comment (AliyevH) [reply Lib/types.py]: No :) I forgot about it

Comment (AliyevH) [reply Lib/types.py]: Fixed

**Review by uriyyo — APPROVED** [review]
LGTM💫

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Thanks for your contribution!

Minor note: in the future please don't force-push as it overrides commit history @AliyevH. Just commit normally, the core devs will squash and merge the PR.

https://devguide.python.org/pullrequest/#quick-guide

**Review by Fidget-Spinner — CHANGES_REQUESTED** [review]
Wait, the docs here need updating too https://docs.python.org/3.10/library/types.html#types.Union.

**Comment by AliyevH** [comment]
> cs he



> LGTM. Thanks for your contribution!
> 
> Minor note: in the future please don't force-push as it overrides commit history @AliyevH. Just commit normally, the core devs will squash and merge the PR.
> 
> https://devguide.python.org/pullrequest/#quick-guide

Got it.  Thanks for support )

File: Misc/NEWS.d/next/Core and Builtins/2021-07-26-15-27-03.bpo-44732.IxObt3.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Rename ``types.Union`` to ``types.UnionType``
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Core and Builtins/2021-07-26-15-27-03.bpo-44732.IxObt3.rst]:
```suggestion
Rename ``types.Union`` to ``types.UnionType``.
```

**Comment by orsenthil** [comment]
As it is already referenced in this ticket, it broke (the stable version of )`sphinx-build` and multiple documentation projects are going to be affected by this change.


This is the trackback from sphinx-build command line.

```
  File "/Users/senthilx/venvs/cpython/bin/sphinx-quickstart", line 5, in <module>
    from sphinx.cmd.quickstart import main
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/senthilx/venvs/cpython/lib/python3.11/site-packages/sphinx/cmd/quickstart.py", line 38, in <module>
    from sphinx.util.console import bold, color_terminal, colorize, nocolor, red  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/senthilx/venvs/cpython/lib/python3.11/site-packages/sphinx/util/__init__.py", line 41, in <module>
    from sphinx.util.typing import PathMatcher
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/senthilx/venvs/cpython/lib/python3.11/site-packages/sphinx/util/typing.py", line 37, in <module>
    from types import Union as types_Union
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: cannot import name 'Union' from 'types' (/Users/senthilx/git/cpython/Lib/types.py)
```

Sphinx has already fixed it in master (https://github.com/sphinx-doc/sphinx/pull/9513) - The fix (aliasing to UnionType) doesn't seem great too.

If we get more complaints, I think, either CPython or Sphinx project need to address this by communication /  guidance.

It was hard to track down this failure with the trace back.

**Comment by Fidget-Spinner** [comment]
@orsenthil that's sad to hear :(. Is there some better place to make such changes known to the wider community?

This change was discussed and agreed upon on python-dev, but that's centered around CPython so I don't think there's much visibility.

**Pull request merged**